src/PolyType/Abstractions/ITypeShape.cs

```diff
@@ -60,6 +60,24 @@ public interface ITypeShape
     /// </remarks>
     /// <exception cref="InvalidOperationException">Thrown when this method is called on an instance that does not represent a generic type.</exception>
     Func<object>? GetAssociatedTypeFactory(Type associatedType);
+
+    /// <summary>
+    /// Gets the shape for a type associated to this property's declared <see cref="Type"/>,
+    /// as captured in <see cref="TypeShapeAttribute.AssociatedShapes"/> or
+    /// <see cref="TypeShapeExtensionAttribute.AssociatedShapes"/>.
+    /// </summary>
+    /// <param name="associatedType">
+    /// The associated type (which must be one found in the <see cref="TypeShapeAttribute.AssociatedTypes"/> property).
+    /// If the associated type is a generic type definition, the type arguments used on this shape's <see cref="Type"/>
+    /// will be used to close the associated generic type.
+    /// </param>
+    /// <returns>The shape for the associated type, or <see langword="null" /> if no shape for the associated type is available.</returns>
+    /// <remarks>
+    /// <see cref="ReflectionProvider.ReflectionTypeShapeProvider"/> can produce the shape on demand without any <see cref="TypeShapeAttribute.AssociatedShapes"/>,
+    /// while <see cref="SourceGenModel.SourceGenTypeShapeProvider"/> is expected to only produce the shape that was explicitly requested via attribute.
+    /// </remarks>
+    /// <exception cref="InvalidOperationException">Thrown when this method is called on an instance that does not represent a generic type.</exception>
+    ITypeShape? GetAssociatedTypeShape(Type associatedType);
```

Comment:
I feel that this method could be unified with the existing `GetAssociatedTypeFactory`. `ITypeShape` already exposes default constructors where available (although it takes a bit of digging with visitors to extract it from the cases that support it).

Would it make sense to expose a `Func<object> GetDefaultConstructor();` method on the `ITypeShape` object directly? This could either be implemented by the interface directly or exposed as an extension method that applies a visitor. In that case `AssociatedTypeRequirements` would morph into an enum that controls the degree/depth of shape generation for a given type: e.g. `DefaultConstructorOnly`, `Transitive`, or possible other modes we can come up with in the future.

Reply:
I can see that.
But I wonder how much more complicated it could potentially make visitors if they have to account for partially complete shapes.

Reply:
I'm sketching this out over here: https://github.com/AArnott/PolyType/pull/1
Your feedback is appreciated